### PR TITLE
Update to use latest etcd client library

### PR DESCRIPTION
**Pull request by timothysc**
This is the update for the new client library after all the storage refactoring.  There is still an error I'm chasing down in the cache, but everything else appears green.  

Final implementation for #11962

@wojtek-t & @jdef for review 

/cc @liggitt

### CODE REVIEW

File: contrib/mesos/pkg/scheduler/service/service.go [hunk 1]
```diff
@@ -629,16 +629,14 @@ func validateLeadershipTransition(desired, current string) {
 }
 
 // hacked from https://github.com/GoogleCloudPlatform/kubernetes/blob/release-0.14/cmd/kube-apiserver/app/server.go
-func newEtcd(etcdConfigFile string, etcdServerList []string) (client *etcd.Client, err error) {
-	if etcdConfigFile != "" {
-		client, err = etcd.NewClientFromFile(etcdConfigFile)
-	} else {
-		client = etcd.NewClient(etcdServerList)
+func newEtcd(etcdServerList []string) (etcd.Client, error) {
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service.go]: looks like `etcdConfigFile` should be completely removed as a state var (and flag) from `SchedulerServer`

File: cmd/integration/integration.go [hunk 2]
```diff
@@ -93,17 +94,24 @@ func (h *delegateHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
-	servers := []string{}
-	glog.Infof("Creating etcd client pointing to %v", servers)
-
 	handler := delegateHandler{}
 	apiServer := httptest.NewServer(&handler)
 
-	etcdClient := etcd.NewClient(servers)
+	cfg := etcd.Config{
+		Endpoints: []string{"http://127.0.0.1:4001"},
+		Transport: etcd.DefaultTransport,
```
Comment (jdef) [inline cmd/integration/integration.go]: not super familiar with the new etcd api -- why is `Transport` assigned to `etcd.DefaultTransport` for some of these configs and not for others?

File: contrib/mesos/pkg/election/etcd_master.go [hunk 3]
```diff
@@ -104,7 +111,11 @@ func (e *etcdMasterElector) extendMaster(path, id string, ttl uint64, res *etcd.
 // returns id, nil if the attempt succeeded
 // returns "", err if an error occurred
 func (e *etcdMasterElector) becomeMaster(path, id string, ttl uint64) (string, error) {
-	_, err := e.etcd.Create(path, id, ttl)
+	opts := etcd.SetOptions{
+		TTL: time.Duration(ttl) * time.Second,
+	}
```
Comment (jdef) [inline contrib/mesos/pkg/election/etcd_master.go]: missing `PrevExist: etcd.PrevNoExist`?

File: pkg/storage/etcd/etcd_watcher.go [hunk 4]
```diff
@@ -390,8 +413,9 @@ func (w *etcdWatcher) Stop() {
...
 	defer w.stopLock.Unlock()
 	// Prevent double channel closes.
-	if !w.stopped {
-		w.stopped = true
-		close(w.userStop)
+	if w.cancel != nil {
+		w.cancel()
+		w.cancel = nil
 	}
+	close(w.userStop)
```
Comment (jdef) [inline pkg/storage/etcd/etcd_watcher.go]: how is a double chan close prevented here?

File: pkg/storage/cacher_test.go [hunk 5]
```diff
@@ -287,7 +287,6 @@ func TestFiltering(t *testing.T) {
 	if err != nil {
 		t.Fatalf("Incorrect resourceVersion: %s", fooCreated.ResourceVersion)
 	}
-	initialVersion--
```
Comment (wojtek-t) [inline pkg/storage/cacher_test.go]: Why did you remove this?

File: pkg/storage/cacher_test.go [hunk 6]
```diff
@@ -299,23 +298,3 @@ func TestFiltering(t *testing.T) {
 	verifyWatchEvent(t, watcher, watch.Modified, podFooPrime)
 	verifyWatchEvent(t, watcher, watch.Deleted, podFooPrime)
 }
-
-/* TODO: So believe it or not... but this test is flakey with the go-etcd client library
```
Comment (wojtek-t) [inline pkg/storage/cacher_test.go]: Also - why do you remove this test? We should uncomment it, I guess.

File: pkg/storage/etcd/etcd_helper.go [hunk 7]
```diff
@@ -56,27 +53,24 @@ func (c *EtcdConfig) GetType() string {
 
 // implements storage.Config
 func (c *EtcdConfig) NewStorage() (storage.Interface, error) {
-	etcdClient := etcd.NewClient(c.ServerList)
-	if etcdClient == nil {
-		return nil, errors.New("Failed to create new etcd client from serverlist")
+	cfg := etcd.Config{
+		Endpoints: c.ServerList,
+		Transport: etcd.DefaultTransport,
+		// TODO: Determine if transport needs optimization
```
Comment (wojtek-t) [inline pkg/storage/etcd/etcd_helper.go]: Can't we leave transport as it was and leave a TODO to revisit it? It would be safer...

File: pkg/storage/etcd/etcd_watcher.go [hunk 8]
```diff
@@ -118,37 +120,59 @@ func newEtcdWatcher(list bool, include includeFunc, filter storage.FilterFunc, e
...
+	defer close(w.etcdIncoming)
 	if resourceVersion == 0 {
-		latest, err := etcdGetInitialWatchState(client, key, w.list, w.etcdIncoming)
+		latest, err := etcdGetInitialWatchState(ctx, client, key, w.list, w.etcdIncoming)
 		if err != nil {
 			w.etcdError <- err
 			return
 		}
-		resourceVersion = latest + 1
+		resourceVersion = latest
```
Comment (wojtek-t) [inline pkg/storage/etcd/etcd_watcher.go]: This is changing the logic - did the logic in client change comparing to go-etcd?

Comment (wojtek-t) [inline pkg/storage/etcd/etcd_watcher.go]: If so - can you please explicitly add a comment  in the storage.Interface, what are the guarantees (I know, we don't have such comment now - it's good time to add it anyway).

Comment (timothysc) [inline pkg/storage/etcd/etcd_watcher.go]: Yup, the logic changed, and I'll add a comment.

File: cmd/integration/integration.go [hunk 9]
```diff
@@ -93,17 +94,24 @@ func (h *delegateHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+	}
+	etcdClient, err := etcd.New(cfg)
+	if err != nil {
+		glog.Fatalf("Error creating etcd client: %v", err)
+	}
+	glog.Infof("Creating etcd client pointing to %v", cfg.Endpoints)
+
+	kAPI := etcd.NewKeysAPI(etcdClient)
```
Comment (liggitt) [inline cmd/integration/integration.go]: nit, but I would assume kapi meant "kubernetes api"... maybe a clearer name?

File: pkg/storage/etcd/etcd_helper.go (hunk 7 above)
Comment (liggitt) [inline pkg/storage/etcd/etcd_helper.go]: +1

File: pkg/storage/etcd/etcd_helper.go [hunk 10]
```diff
@@ -119,7 +114,27 @@ func (h *etcdHelper) Backends(ctx context.Context) []string {
...
+	memberLength := len(members)
+	if 0 == memberLength {
+		return nil
+	}
+
+	mlist := make([]string, memberLength)
+	for idx, member := range members {
+		if len(member.ClientURLs) != 0 {
+			// TODO: Is there a case where this is invalid?
+			mlist[idx] = member.ClientURLs[0]
```
Comment (liggitt) [inline pkg/storage/etcd/etcd_helper.go]: the old GetCluster impl appended all ClientURLs, not just the first

File: pkg/storage/etcd/testing/utils.go [hunk 11]
```diff
@@ -145,9 +145,13 @@ func NewEtcdTestClientServer(t *testing.T) *EtcdTestServer {
 		t.Fatal("Failed to start etcd server error=%v", err)
 		return nil
 	}
-	server.Client = goetcd.NewClient(server.ClientURLs.StringSlice())
-	if server.Client == nil {
-		t.Errorf("Failed to connect to local etcd server")
+	cfg := etcd.Config{
+		Endpoints: server.ClientURLs.StringSlice(),
+		Transport: etcd.DefaultTransport,
```
Comment (liggitt) [inline pkg/storage/etcd/testing/utils.go]: can omit Transport, if nil, etcd.DefaultTransport is used

File: contrib/mesos/pkg/scheduler/service/service.go (hunk 1 above)
Comment (liggitt) [inline contrib/mesos/pkg/scheduler/service/service.go]: were you going to update `newEtcd` in `kube-apiserver/app/server.go`? also, what are we going to do about `etcdConfigFile`? that's currently the only way to connect to a secured etcd

Comment (timothysc) [inline contrib/mesos/pkg/scheduler/service/service.go]: re: etcdConfigFile - it's completely removed in the nwo.

**Comment by wojtek-t** [comment]
@timothysc - I think that this PR should fix the remaining issue in Go 1.5 - see https://github.com/kubernetes/kubernetes/issues/18701 for more details.

Can we prioritize it to have this merged ASAP?

**Comment by timothysc** [comment]
on it.

**Comment by wojtek-t** [comment]
@timothysc - thanks a lot!

File: cmd/integration/integration.go (hunk 2 above)
Comment (timothysc) [inline cmd/integration/integration.go]: I can remove, it only really needs to be commented in a single location in the storage layer really. 

Done.

File: cmd/integration/integration.go (hunk 9 above)
Comment (timothysc) [inline cmd/integration/integration.go]: Done.

File: contrib/mesos/pkg/election/etcd_master.go (hunk 3 above)
Comment (timothysc) [inline contrib/mesos/pkg/election/etcd_master.go]: yup, and done.

File: contrib/mesos/pkg/scheduler/service/service.go (hunk 1 above)
Comment (timothysc) [inline contrib/mesos/pkg/scheduler/service/service.go]: @jdef Done. 

> were you going to update newEtcd in kube-apiserver/app/server.go ?

already done.

File: pkg/storage/etcd/etcd_helper.go (hunk 7 above)
Comment (timothysc) [inline pkg/storage/etcd/etcd_helper.go]: Transport depended on a forked go-etcd client Dial class, which seemed like a hack.  IMHO we should ping @xiang90 to determine optimized path vs. maintaining a fork of the deprecated interface.

Comment (liggitt) [inline pkg/storage/etcd/etcd_helper.go]: agree we should check, but I want to maintain current behavior until that is resolved

File: pkg/storage/etcd/etcd_helper.go (hunk 10 above)
Comment (timothysc) [inline pkg/storage/etcd/etcd_helper.go]: Fixed.

File: pkg/storage/etcd/etcd_watcher.go (hunk 4 above)
Comment (timothysc) [inline pkg/storage/etcd/etcd_watcher.go]: removed old comment as previous logic doesn't apply in new situation.

File: pkg/storage/etcd/testing/utils.go (hunk 11 above)
Comment (timothysc) [inline pkg/storage/etcd/testing/utils.go]: Done.

**Comment by timothysc** [comment]
@wojtek-t I think I'll need to ping you tomorrow on the cache tomorrow, there is some subtlety that I'm missing that is causing an issue there.

**Comment by wojtek-t** [comment]
@timothysc - sure

**Comment by wojtek-t** [comment]
@timothysc - can you please rebase this PR - it will be much simpler for me to patch it and debug the cacher problem

**Comment by wojtek-t** [comment]
@timothysc - cacher_test.go - line 205 (initialVersion--) should be removed

This solves the problem (at least when I'm running it locally).

File: contrib/mesos/pkg/scheduler/service/service.go [hunk 12]
```diff
@@ -689,10 +685,11 @@ func (s *SchedulerServer) bootstrap(hks hyperkube.Interface, sc *schedcfg.Config
 	// (1) the generic config store is available for the FrameworkId storage
 	// (2) the generic master election is provided by the apiserver
 	// Compare docs/proposals/high-availability.md
-	etcdClient, err := newEtcd(s.etcdConfigFile, s.etcdServerList)
+	etcdClient, err := newEtcd(s.etcdServerList)
 	if err != nil {
 		log.Fatalf("misconfigured etcd: %v", err)
 	}
+	kAPI := etcd.NewKeysAPI(etcdClient)
```
Comment (wojtek-t) [inline contrib/mesos/pkg/scheduler/service/service.go]: s/kAPI/keysAPI/ ?

File: pkg/storage/cacher_test.go (hunk 5 above)
Comment (timothysc) [inline pkg/storage/cacher_test.go]: RV shift.

File: contrib/mesos/pkg/scheduler/service/service.go [hunk 13]
```diff
@@ -741,7 +738,7 @@ func (s *SchedulerServer) bootstrap(hks hyperkube.Interface, sc *schedcfg.Config
 		LookupNode:        lookupNode,
 		StoreFrameworkId: func(id string) {
 			// TODO(jdef): port FrameworkId store to generic Kubernetes config store as soon as available
-			_, err := etcdClient.Set(meta.FrameworkIDKey, id, uint64(s.failoverTimeout))
+			_, err := kAPI.Set(context.TODO(), meta.FrameworkIDKey, id, &etcd.SetOptions{TTL: time.Duration(s.failoverTimeout) * time.Second})
```
Comment (wojtek-t) [inline contrib/mesos/pkg/scheduler/service/service.go]: I think PrevExist should be set here in options.

File: pkg/storage/cacher_test.go (hunk 6 above)
Comment (timothysc) [inline pkg/storage/cacher_test.go]: A replica of this test exists at the etcd layer and is enabled now.  I removed b/c it seemed redundant.

File: pkg/storage/cacher_test.go [hunk 14]
```diff
@@ -202,7 +202,6 @@ func TestWatch(t *testing.T) {
 	if err != nil {
 		t.Fatalf("Incorrect resourceVersion: %s", fooCreated.ResourceVersion)
 	}
```
Comment (wojtek-t) [inline pkg/storage/cacher_test.go]: Since you are no longer changing initialVersion, please remove parsing it and use fooCreated.ResourceVersion instead

File: pkg/storage/cacher_test.go [hunk 15]
```diff
@@ -287,7 +286,6 @@ func TestFiltering(t *testing.T) {
...
```
Comment (wojtek-t) [inline pkg/storage/cacher_test.go]: Since you are no longer changing initialVersion, please remove parsing it and use fooCreated.ResourceVersion instead

File: pkg/storage/cacher_test.go (hunk 6 above)
Comment (wojtek-t) [inline pkg/storage/cacher_test.go]: ping

File: pkg/storage/etcd/etcd_helper.go [hunk 16]
```diff
@@ -56,27 +53,23 @@ func (c *EtcdConfig) GetType() string {
 
 // implements storage.Config
 func (c *EtcdConfig) NewStorage() (storage.Interface, error) {
-	etcdClient := etcd.NewClient(c.ServerList)
-	if etcdClient == nil {
-		return nil, errors.New("Failed to create new etcd client from serverlist")
+	cfg := etcd.Config{
+		Endpoints: c.ServerList,
+		// Transport: TODO: Determine if transport needs optimization
```
Comment (timothysc) [inline pkg/storage/etcd/etcd_helper.go]: I'm inclined to profile 1st, then optimize here vs. preserving what was a hack.

Comment (wojtek-t) [inline pkg/storage/etcd/etcd_helper.go]: Can we set transport the same way as before (i.e MaxIdleConnsPerHost to 500)?

File: pkg/storage/etcd/etcd_watcher.go [hunk 17]
```diff
@@ -136,37 +138,59 @@ func newEtcdWatcher(list bool, include includeFunc, filter storage.FilterFunc, e
...
+	opts := etcd.WatcherOptions{
+		Recursive:  w.list,
+		AfterIndex: resourceVersion,
+	}
+	watcher := client.Watcher(key, &opts)
+	w.ctx, w.cancel = context.WithCancel(ctx)
+
+	for {
+		resp, err := watcher.Next(w.ctx)
+		if resp != nil {
```
Comment (wojtek-t) [inline pkg/storage/etcd/etcd_watcher.go]: This is slightly strange.
Can we do it like this:

if err != nil {
  w.etcdError <- err
  return
}
w.etcdIncoming <- resp

I think this is clearer

File: pkg/storage/etcd/etcd_watcher.go (hunk 4 above)
Comment (wojtek-t) [inline pkg/storage/etcd/etcd_watcher.go]: But we can potentially still close the channel more than once now. It doesn't seem to be safe.

File: test/integration/etcd_tools_test.go [hunk 18]
```diff
@@ -26,24 +26,26 @@ import (
...
 	"k8s.io/kubernetes/test/integration/framework"
 
+	etcd "github.com/coreos/etcd/client"
 	"golang.org/x/net/context"
 )
 
 func TestSet(t *testing.T) {
 	client := framework.NewEtcdClient()
-	etcdStorage := etcd.NewEtcdStorage(client, testapi.Default.Codec(), "")
+	kAPI := etcd.NewKeysAPI(client)
```
Comment (wojtek-t) [inline test/integration/etcd_tools_test.go]: s/kAPI/keysAPI/

File: test/integration/etcd_tools_test.go [hunk 19]
```diff
@@ -140,11 +144,12 @@ func TestWriteTTL(t *testing.T) {
 
 func TestWatch(t *testing.T) {
 	client := framework.NewEtcdClient()
-	etcdStorage := etcd.NewEtcdStorage(client, testapi.Default.Codec(), etcdtest.PathPrefix())
+	kAPI := etcd.NewKeysAPI(client)
```
Comment (wojtek-t) [inline test/integration/etcd_tools_test.go]: s/kAPI/keysAPI/

File: test/integration/framework/etcd_utils.go [hunk 20]
```diff
@@ -68,13 +77,13 @@ func WithEtcdKey(f func(string)) {
 // of the test run.
 func DeleteAllEtcdKeys() {
 	glog.Infof("Deleting all etcd keys")
-	client := NewEtcdClient()
-	keys, err := client.Get("/", false, false)
+	kAPI := etcd.NewKeysAPI(NewEtcdClient())
```
Comment (wojtek-t) [inline test/integration/framework/etcd_utils.go]: s/kAPI/keysAPI/

File: pkg/storage/etcd/etcd_helper.go (hunk 16 above)
Comment (liggitt) [inline pkg/storage/etcd/etcd_helper.go]: yes, please leave the bumped MaxIdleConnsPerHost

File: cmd/integration/integration.go [hunk 21]
```diff
@@ -63,9 +63,10 @@ import (
 	"k8s.io/kubernetes/test/integration"
 	"k8s.io/kubernetes/test/integration/framework"
 
-	"github.com/coreos/go-etcd/etcd"
+	etcd "github.com/coreos/etcd/client"
```
Comment (wojtek-t) [inline cmd/integration/integration.go]: Can you please add a comment to Watch() and WatchList() functions in:
pkg/storage/interface.go
about what resourceVersion semantics is (that we start right after the passed version).

File: contrib/mesos/pkg/scheduler/service/service.go (hunk 12 above)
Comment (timothysc) [inline contrib/mesos/pkg/scheduler/service/service.go]: Done.. I may have accidentally nixed a some changes in trying to share debugging.

File: contrib/mesos/pkg/scheduler/service/service.go (hunk 13 above)
Comment (timothysc) [inline contrib/mesos/pkg/scheduler/service/service.go]: It wasn't previously... The behavior is preserved from the previous code.

File: pkg/storage/cacher_test.go (hunk 14 above)
Comment (timothysc) [inline pkg/storage/cacher_test.go]: Done.

File: pkg/storage/cacher_test.go (hunk 15 above)
Comment (timothysc) [inline pkg/storage/cacher_test.go]: Done.

File: pkg/storage/etcd/etcd_helper.go (hunk 16 above)
Comment (timothysc) [inline pkg/storage/etcd/etcd_helper.go]: Done.

File: pkg/storage/etcd/etcd_watcher.go (hunk 17 above)
Comment (timothysc) [inline pkg/storage/etcd/etcd_watcher.go]: I put the extra logic in as a guard in case of API misbehavior originally, but tests are green on change. 

Done.

File: test/integration/etcd_tools_test.go (hunk 18 above)
Comment (timothysc) [inline test/integration/etcd_tools_test.go]: Done.

File: test/integration/etcd_tools_test.go (hunk 19 above)
Comment (timothysc) [inline test/integration/etcd_tools_test.go]: Done.

File: test/integration/framework/etcd_utils.go (hunk 20 above)
Comment (timothysc) [inline test/integration/framework/etcd_utils.go]: Done.

File: cmd/integration/integration.go (hunk 21 above)
Comment (timothysc) [inline cmd/integration/integration.go]: Done.

**Comment by timothysc** [comment]
@wojtek-t comments addressed. PTAL.

File: pkg/storage/etcd/etcd_watcher.go (hunk 4 above)
Comment (timothysc) [inline pkg/storage/etcd/etcd_watcher.go]: fixed.

File: pkg/storage/cacher_test.go (hunk 6 above)
Comment (wojtek-t) [inline pkg/storage/cacher_test.go]: got it - thanks!

**Comment by wojtek-t** [comment]
@timothysc - thanks a lot for this PR

LGTM

**Comment by wojtek-t** [comment]
[Testing before manually merging]

@k8s-bot test this please

**Comment by wojtek-t** [comment]
unit tests passed, e2e flake

@k8s-bot test this please

**Comment by wojtek-t** [comment]
infrastructure unit-test failure:
 > git checkout -f b43af14b26801071c65ee65bab22f5247840dd15
FATAL: Could not checkout b43af14b26801071c65ee65bab22f5247840dd15

@k8s-bot unit test this please

**Comment by wojtek-t** [comment]
The same again:
 > git checkout -f b43af14b26801071c65ee65bab22f5247840dd15
FATAL: Could not checkout b43af14b26801071c65ee65bab22f5247840dd15

@k8s-bot test this please

**Comment by wojtek-t** [comment]
@k8s-bot e2e test this please

**Comment by wojtek-t** [comment]
All tests are passing - merging manually, to unblock effort to update to Go 1.5

**Comment by timothysc** [comment]
Awesome!

**Comment by ixdy** [comment]
I think this PR started causing several etcd unit tests to data race or timeout, as linked in the above issues.

File: pkg/storage/etcd/etcd_helper.go [hunk 22]
```diff
@@ -56,27 +55,32 @@ func (c *EtcdConfig) GetType() string {
...
-		return nil, errors.New("Failed to create new etcd client from serverlist")
-	}
-	transport := &http.Transport{
-		Dial: forked.Dial,
-		TLSClientConfig: &tls.Config{
-			InsecureSkipVerify: true,
+	cfg := etcd.Config{
+		Endpoints: c.ServerList,
+		// TODO: Determine if transport needs optimization
+		Transport: &http.Transport{
```
Comment (jdef) [inline pkg/storage/etcd/etcd_helper.go]: why not use `util.SetTransportDefaults` here?

Comment (wojtek-t) [inline pkg/storage/etcd/etcd_helper.go]: We are overwriting some things (e.g. MaxIdleConnsPerHost) - it was crucial for performance reasons.

Comment (liggitt) [inline pkg/storage/etcd/etcd_helper.go]: You can still do that with SetTransportDefaults… only set MaxIdle… and the defaulting func will set all the other defaults

Comment (timothysc) [inline pkg/storage/etcd/etcd_helper.go]: The defaults are explicitly specified so others will know the values on inspection, vs. hunt and peck.

**Comment by jdef** [comment]
xref #18153

File: contrib/mesos/pkg/scheduler/service/service.go (hunk 1 above)
Comment (bprashanth) [inline contrib/mesos/pkg/scheduler/service/service.go]: so what's the recommended way to connect securely to etcd right now? we should surface that somewhere if we haven't already, I see a good amount of confusion on slack.

**Comment by streamnsight** [comment]
@timothysc 
@jdef 
@wojtek-t 

I am running into this issue:

https://github.com/kubernetes/kubernetes/issues/24578

trying to run apiserver with etcd TLS.

TLS worked fine with v1.1.2 I am trying to upgrade from.
I could get into the apiserver container and cat the files apiserver is complaining about not finding, so I don't understand what is wrong.

**Comment by streamnsight** [comment]
same problem with 1.2.1

**Comment by streamnsight** [comment]
in using the etcd client before I had to have soemthign like 

```
var cfg store.Config

	if caCert != "" && clientCert != "" && clientKey != "" {
		var tlsInfo = transport.TLSInfo{
			CAFile:   caCert,
			CertFile: clientCert,
			KeyFile:  clientKey,
			TrustedCAFile: caCert,
			ClientCertAuth: true,
		}
		t, err := transport.NewTransport(tlsInfo, 10 * time.Second)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}

		cfg = store.Config{
			ConnectionTimeout: 10 * time.Second,
			TLS: t.TLSClientConfig,
		}
	} else {
		cfg = store.Config{
				ConnectionTimeout: 10 * time.Second,
			}
	}
```

**Comment by timothysc** [comment]
@streamnsight The fix is here #21535, but it changes the default initialization.  I believe the release note denotes as well.

**Comment by streamnsight** [comment]
@timothysc   
Sorry Tim but I don't get it: 
I don't see anything in the release notes, and I don't understand the 'fix': what I see is a PR but no explanation about how I'm now suppose to configure secured etcd properly. 
If i go by the docs it doesn't work as intended.

Could you please explain what I'm supposed to do or point me to the proper documentation.
Thanks

**Comment by liggitt** [comment]
the updates needed for secured etcd were in the [v1.2.0 release notes](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG.md/#action-required-1) in the "action required" section: 

```
Running against a secured etcd requires these flags to be passed to kube-apiserver (instead of --etcd-config):
    --etcd-certfile, --etcd-keyfile (if using client cert auth)
    --etcd-cafile (if not using system roots)
```

**Comment by streamnsight** [comment]
@liggitt  Thanks for pointing this out.

**Comment by streamnsight** [comment]
FYI the problem I was having was having quotes around the file names in the etcd config.
This is a pretty stupid mistake, but I thought yaml supported both quoted or not quoted string, so just wanted to point it out here.
